build_retrieval_dataset.py

```diff
@@ -0,0 +1,389 @@
+    if blur_strgth == "low":
+        inp_pic_grey_blurred = cv.GaussianBlur(inp_pic_grey, (3,3), 1, 1)
+    elif blur_strgth == "high":
+        inp_pic_grey_blurred = cv.GaussianBlur(inp_pic_grey, (9,9), 2, 2)
+    #cv.imshow("grey_blurred", inp_pic_grey_blurred)
+    # HoughCircles(image, method, dp, minDist, circles=None, param1=None, param2=None, minRadius=None, maxRadius=None)
+    circles = cv.HoughCircles(inp_pic_grey_blurred, cv.HOUGH_GRADIENT, hough_dp, 20, minRadius, maxRadius) #minDist = 20
+    if (circles is None):
+        print("No circles found.")
+        raise Exception("No circles found.")
+    elif circles.shape == (4,1):
+        # print("More than one circle found.")
+        # For some images, the detection fails and openCV returns a shape of (4,1).
+        # I cannot find this behaviour in the documentation, so maybe it is a bug
+        # Best fix so far: guess the circle's position
+        y, x = inp_pic_grey.shape[:2]
+        return int(x/2), int(y/2), int(min(x,y) / 2 * 0.95)
+    else:
+        circles = np.round(circles[0, :].astype("int")) # rounding coordinates to integer values
+        x_ctr, y_ctr, r = circles[0]
+        #cv.circle(inp_pic, (125, 125), r, color=(0, 0, 0), thickness=4, lineType=8, shift=0)
+        #cv.imshow('circle in inp_pic', inp_pic)
+        # print("1 circle found. radius: ", r, ", center coordinate: (", x_ctr, ",", y_ctr, ")")
+        return x_ctr, y_ctr, r
+
+
+def create_masks(x_ctr, y_ctr, r, x=250, y=250):
+    """
+    Creates mask (and inverted mask) with specific size (y, x) and white (black) circle (x_ctr, y_ctr, r) in it.
+    :param x: width of mask
+    :param y: height of mask
+    :param x_ctr: x-coordinate of circle center
+    :param y_ctr: y-coordinate of circle center
+    :param r: radius of circle
+    :return: mask (white filled circle) and mask_inv (black filled circle)
+    """
+    mask = np.zeros((y, x, 3), np.uint8)
+    center = (x_ctr, y_ctr)
+    # cv.circle(img, center, radius, color[, thickness[, lineType[, shift]]]) -> img
+    cv.circle(mask, center, r, color=(255, 255, 255), thickness=-1, lineType=8, shift=0) # thickness=-1 => filled circle
+    mask = cv.cvtColor(mask, cv.COLOR_BGR2GRAY)
+    mask_inv = cv.bitwise_not(mask)
+    return mask, mask_inv
+
+
+def extract_coin(inp_pic, mask):
+    """
+    Takes an image and extracts the coin with respect to the given coin-mask.
+    :param inp_pic: input picture
+    :param mask: coin-mask
+    :return: output picture of the coin without a background
+    """
+    out_pic = cv.bitwise_and(inp_pic, inp_pic, mask=mask)
+    return out_pic
+
+
+def resize_pic(inp_pic, x=64, y=64):
+    """
+    Resize a 250x250 input picture to 64x64 output picture.
+    :param inp_pic: input picture
+    :param x: x, width
+    :param y: y, height
+    :return: output picture
+    """
+    r = x / y
+    dim = (x, int(x * r))
+    out_pic = cv.resize(inp_pic, dim, interpolation=cv.INTER_AREA)
```

Comment:
Sorry, I don't get why `dim` is computed in this way, `out_pic = cv.resize(inp_pic, (x,y), interpolation=cv.INTER_AREA)` should be fine.

Reply:
Actually, I think this might even be a bug.
E.g., consider `x=128`, `y=64`. This would result in `r = 128 / 64 = 2`, which would then result in an image size of `(x, 2x)`, but unless i'm mistaken, it should rather be `(x, 1/2 x)`. @roman-staehle I don't recall writing this function, so it is probably yours :sweat_smile:  Are we overlooking something, or can we get rid of the `r`?

Reply:
was general thought as scaling from 250x250 to a variable chosen dim size depending on finding suiting y to x without loosing the scale ratio. hard fix like recommended:
`out_pic = cv.resize(inp_pic, (y, x), interpolation=cv.INTER_AREA)`
(y: row, x: col)

Reply:
Good, if we center-crop first to an equal amount of pixels in width and height we should not worry about the scale ratio.